OCDEP 101- fix some inconsistencies in how we handle datetimes

jamesturk opened the pull request:
deals with opencivicdata/python-opencivicdata#87 and some other nagging time-related issues in the hopes of fixing a few at once 

cc @fgregg @jpmckinney

jamesturk committed:
add datetime proposal

jamesturk committed:
fix some formatting issues

jpmckinney commented on proposals/drafts/datetimes.rst:
```diff
@@ -0,0 +1,108 @@
+* Post.{start_date,end_date}
+
+"Fuzzy Datetime Field"
+
```
Remove new line to have definition list style instead of blockquote.

jpmckinney commented on proposals/drafts/datetimes.rst:
```diff
@@ -0,0 +1,108 @@
+
+"Fuzzy Datetime Field"
+
+    For VoteEvents sometimes the time is important too, so we extended the aove field to 19 characters, allowing an additional inclusion of time in HH:MM:SS.
```
aove -> above

jpmckinney commented on proposals/drafts/datetimes.rst:
```diff
@@ -0,0 +1,108 @@
+VoteEvent's special case of fuzzy time is a problem in two ways:
+
+1) it can have a time but lacks a timezone
+2) it is misnamed, ending in _date while allowing a time to be stored
```
Other specifications consistently use `date` whether it's a date or datetime. `time` suffixes are popular within software, kind of like the `_i` or `_s` suffixes popular in C programming to flag a variable's type in the variable name. But `date` is not a misnaming; it's a choice that makes the field name about the semantics of the field, rather than an indicator of its type.

jpmckinney commented on proposals/drafts/datetimes.rst:
```diff
@@ -0,0 +1,108 @@
+* Define that time is always stored in UTC, but this would be more
+  error prone than being explicit.
+
+2) To address #2, rename VoteEvent.start_date,end_date to start_time,end_time to match Event and be more clear.
```
Renaming would break with Popolo (and the vast majority of specifications with time fields, like Schema.org, Dublin Core, etc.). I explain why 2) is not a problem above.

jpmckinney commented on proposals/drafts/datetimes.rst:
```diff
@@ -0,0 +1,108 @@
+1) To address #1 and #5: add timezone to "fuzzy datetime" and bring the full format in line w/ ISO8601, changing the format from:
+        YYYY[[[-MM]-DD] HH:MM:SS]
+            to 
+        YYYY[[[-MM]-DD]THH:MM:SS(Z|+XX:YY)]
```
Popolo's validation allows one of:

* YYYY[-MM[-DD]]
* YYYY-MM-DD[THH[:MM[:SS]]Z]

Time zone support hadn't been added because it had never been requested, but I can make that change easily.

jamesturk committed:
update proposal based on feedback

this withdraws the name changes to *time, but does suggest that the existing single use of *time is changed to *date for consistency

jpmckinney commented on proposals/drafts/datetimes.rst:
```diff
@@ -0,0 +1,104 @@
+
+For the most part this is OK, and we're fairly consistent.  Most uses of fuzzy date align with the goals, but in a few cases it seems like we've made some mistakes:
+
+1) VoteEvent and Event have very similar start/end times with completely different semantics.  VoteEvent's special case of fuzzy time can have a time but lacks a timezone, while Event's fields are named start_time/end_time and use a DateTime object, the only place one is used (requiring more precision than we're guaranteed to have).
```
I think the semantics/meaning are the same (in Popolo, VoteEvent inherits start/end date properties from Event); it's the types that are completely different.

jpmckinney commented on proposals/drafts/datetimes.rst:
```diff
@@ -0,0 +1,104 @@
+We'd make the following changes:
+
+1) To address #1 and #2: add timezone to "fuzzy datetime" and bring the full format in line w/ ISO8601, changing the format from:
+        YYYY[[[-MM]-DD] HH:MM:SS]
```
It looks like it's actually: `[YYYY][-MM[-DD]][ HH:MM:SS]` at present: https://github.com/opencivicdata/pupa/blob/master/pupa/scrape/schemas/common.py

jpmckinney commented on proposals/drafts/datetimes.rst:
```diff
@@ -0,0 +1,104 @@
+1) To address #1 and #2: add timezone to "fuzzy datetime" and bring the full format in line w/ ISO8601, changing the format from:
+        YYYY[[[-MM]-DD] HH:MM:SS]
+            to 
+        YYYY[[[-MM]-DD]THH:MM:SS(Z|+XX:YY)]
```
This is the pattern in Popolo (all whitespace is only for readability):

```
^
[0-9]{4}
(
  (-[0-9]{2}){0,2} |
  (-[0-9]{2}){2} T [0-9]{2}(:[0-9]{2}){0,2} (Z | [+-] [0-9]{2}(:[0-9]{2})?)
)
$
```

In the above notation, it'd be like: `YYYY[-MM[-DD[THH[:MM[:SS]](Z|(+|-)HH[:MM])]]]`

What look like unnecessary alternatives in the regex prevent weird cases like `2010T10Z`: ISO8601 doesn't allow setting the hour without the day set<del>, and doesn't allow setting the timezone without the seconds set</del>.

*Update:* Actually, you can set timezone with just minute or hour. Updated.

jamesturk replied:
👍 I'll make sure we adopt the same regex & update this 

there should be a [+-] in there too right?

jpmckinney replied:
Yup, added in an edit :)

jpmckinney reviewed (commented):
The only substantive change in this review is the suggestion of a specific regex once it comes to implementing fuzzy datetimes. Looks good.

jamesturk committed:
update with latest changes to OCDEP101

- assign a number (staking claim to 1xx for amending earlier OCDEPs)
- adopt James McKinney's regex for fuzzy datetimes
- add recommendation to drop ``timezone`` from Event

jamesturk committed:
fix typo

jamesturk committed:
will be #101
